Merge master into nullable branch

safern opened the pull request:
This time there were some merge conflicts in `AssemblyName`, `AssemblyLoadContext`, `TimerQueue`, `RuntimeAssembly` and `Environment.Win32` -- looking at those files in the merge commit would be appreciated to just make sure it looks good and that I didn't miss anything during the merge conflicts.

Commit:
Delete condition that was previously checked with noway_assert.

Commit:
Destroy the old tree in `optAssertionProp_Update`.

Currently compiletion will fail with an assert because of `optPrepareTreeForReplacement`.

Commit:
Clean optVNConstantPropOnTree.

Create a new tree with const val and attach side-effects to it.
Extract common code and delete returns in the of the switch.

Commit:
Clarify some unsupported cases in `optVNConstantPropOnTree`.

Commit:
Delete the asignment that was added as a workaround.

adityamandaleeka committed:
Enable CLR write watch for ARM64 Windows.

mjsabby committed:
Add Large pages support in GC

briansull committed:
Block the hoisting of TYP_STRUCT rvalues in loop hoisting
Added test case GitHub_23739.cs

janvorli committed:
Remove Unix CPU groups emulation

This change removes CPU groups emulation from Unix PAL and modifies the
GC and thread pool code accordingly.

jkotas committed:
Update CoreRT for changes in shared partition

Signed-off-by: dotnet-bot <dotnet-bot@microsoft.com>

safern committed:
Add back using to EventSourceException

Signed-off-by: dotnet-bot <dotnet-bot@microsoft.com>

wtgodbe committed:
Re-update branding to preview5

Commit:
Diable CoreFX System.IO.Tests.dll on windows arm64.

Commit:
Remove the MultiFileSec option from EventPipe. (#23777)

This option was a pseudo mechanism to fake "streaming" events out-of-proc.
The idea was to have EventPipe creating files every N seconds, with event data up to that point. Thus, external processes could read these files in an attempt to get "read-time" data.
Now, we actually have streaming of event through IPC channels, so this option is not needed.

wtgodbe committed:
Merge pull request #23813 from wtgodbe/PrevAgain

Re-update branding to preview5

janvorli committed:
Merge pull request #23251 from mjsabby/largePagesInGC

Add Large pages support in GC

Commit:
Merge pull request #23814 from sandreenko/disableSystemIOTestsCorefx

Diable CoreFX System.IO.Tests.dll on windows arm64.

adityamandaleeka committed:
Merge pull request #23752 from adityamandaleeka/arm64_clrww

Enable CLR write watch for ARM64 Windows

janvorli committed:
Fix several issues

* Fix build on OSX and Linux machines without NUMA installed - there were
couple of places where I was missing ifdefs
* Fix bug in nodeMaskLength computation
* Remove testing change in eeconfig.cpp that has leaked into the PR
* Fix GCToOSInterface::GetTotalProcessorCount for embedded GC to return
all processors on the system, not just the ones enabled for the current
process.

briansull committed:
Merge pull request #23739 from briansull/struct-cse

Block the hoisting of TYP_STRUCT rvalues in loop hoisting

VSadov committed:
Remove exclusion for System.Threading.Tasks.Tests (#23818)

* Update exclusion for System.Threading.Tasks.Tests

There is a race in the test, which causes failures in various stress runs. And that is not Linux specific.

The failure is addressed in https://github.com/dotnet/corefx/pull/36645  
This is to update the exclusion lists to make it not just for Linux and to link to the fix PR, so that it could be tracked when it makes it into coreclr.

* Update corefx_test_exclusions.txt

* Update corefx_test_exclusions.txt

stephentoub committed:
Improve RuntimeHelpers.GetSubArray (#23829)

* Improve RuntimeHelpers.GetSubArray

This change does three things.

First, it fixes `GetSubArray` to work when the supplied array is actually a `U[]` where `U : T`.  Currently this case ends up throwing an exception inside of span, which doesn't like working with arrays covariantly.

Second, it fixes argument validation so that we throw an ArgumentNullException if the input array is null rather than NullReferenceException.

Third, it improves the performance of `GetSubArray` for the 95% common case where either `T` is a value type or the type of the array matches the `T` type specified.

* Only use `Array.Empty<T>` when `typeof(T[]) == array.GetType()`

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20190408.16 (#23836)

- Microsoft.NETCore.App - 3.0.0-preview5-27608-16

dotnet-maestro-bot committed:
Update CoreClr to preview5-27609-71 (#23809)

echesakov committed:
Group XUnitWrapper.dll into the same Payloads when submit to Helix (#23476)

* Add DisplayNameAttribute to each Fact

* Split files under bin/tests folder into groups and specify the corresponding *.XUnitWrapper.dll

* Group some of the *.XUnitWrapper.dll files into PayloadGroups and split others into different PayloadGroups if they have more than one TestGroup defined

* Add "TestGroup" trait to XUnitWrapper facts

* Add Microsoft.DotNet.XUnitConsoleRunner to eng\Versions.props

* Overwrite xunit.console.dll with the version coming from Microsoft.DotNet.XUnitConsoleRunner

echesakov committed:
Unconditionally set CrossGenTest to false for readytorun/tests/mainv1 mainv2 (#23826)

Commit:
Exclude PInvokes declared on other modules. We don't yet encode crossmodule references (#23828)

* Exclude PInvokes declared on other modules. We don't yet encode cross module references

AaronRobinsonMSFT committed:
Alter CCW wrapping semantics (#23709)

* Update CCW semantics to not unwrap when a managed COM server was activated
  from a managed COM client. This is a functional change from .NET Framework.

* Add support for CoreShim to "attach" to the existing CLR instance when
 running from a CoreRun scenario.

* Add testing for NET COM client activating a NET COM server

jkoritzinsky committed:
Don't denormalize enum return values on instance methods on x86. (#23816)

* Don't denormalize enum return values on instance methods on x86.

* Reuse closedElemType.

gregg-miskelly committed:
Add 'SetNotificationForWaitCompletion' to ILLinkTrim.xml (#23822)

SetNotificationForWaitCompletion is meant to be func-eval'ed by the debugger to enable async step out. But this method was being removed by the IL linker. This adds it back.

nxtn committed:
Remove dlclose call to unload CoreCLR (#23863)

sywhang committed:
Expose new EventCounter APIs (#23808)

* rename BaseCounter to DiagnosticCounter

* Change MetaData->Metadata

* Make EventSource and Name a property for counter classes

* Make the counter APIs public

* fix build errors

* Change float to double

* Few cleanups, fix test

* fix GetMetadataString

* PR feedback

* More PR feedback

Commit:
Fix newlines for  ilegal trees in `gtDispTree`.

Commit:
Change optExtractSideEffList to optExtractSideEffList.

janvorli committed:
Fix RecycledLists size on Unix

The list size was set to g_SystemInfo.dwNumberOfProcessors which is a
number of processors the current process is allowed to run on, but not
the total number of processors in the system. Fixed to use
PAL_GetTotalCpuCount.

Also revert a change to the mbind node mask length computation I've
incorrectly made in my last commit and make it clear that the value is
a number of used bits in the node mask, which is the highest numa node
plus 1. And finally, re-reading the mbind doc, I've found that the
maxnode parameter is in fact "number of nodes" in the mask, so fixing
that too.

noahfalk committed:
Update profiling-api-status.md

Clarifying that profiler attach/detach are a known gap

janvorli committed:
Fix few issues discovered by testing

jaredpar committed:
Update Windows instructions (#23841)

The instructions for CMake failed to list 3.14.1 when using Visual Studio 2019.

jkotas committed:
Disable nullability-related warnings

Signed-off-by: dotnet-bot <dotnet-bot@microsoft.com>

dotnet-maestro-bot committed:
Update CoreClr to preview5-27610-71 (#23859)

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20190409.17 (#23865)

- Microsoft.NETCore.App - 3.0.0-preview5-27609-17

tannergooding committed:
Tweaking some APIs according to user feedback. (#23857)

* Updating the Avx2.ConvertToVector256UInt APIs to return a signed type (matching the native signature and Sse41 APIs).

* Adding explicit overloads for `Sse41.ConvertToVector128Int*` and `Avx2.ConvertToVector256Int*` that take a pointer

* Updating the word ordering for the Compare/CompareScalar methods.

tarekgh committed:
Update Range.GetOffsetAndLength (#23855)

* Update Range.GetOffsetAndLength

Make it return offset and length tuple instead of the old Type Range.OffsetAndLength

* Fix the return Tuple fields names

Commit:
Update the comment.

briansull committed:
Improved JitDump
- Prints the bbNatLoopNum when dumping the BasicBlocks
- Refactor the GenTree node dump to use gtDispCommonEndLine:
//     Utility function that prints the following node information
//       1: The associated zero field sequence (if any)
//       2. The register assigned to this node (if any)
//       2. The value number assigned (if any)
//       3. A newline character

briansull committed:
jit-format

jkoritzinsky committed:
Check for a null import descriptor when fixing up vtables. (#23825)

A fragile ngened image doesn't have an import descriptor directory in its PE image. Without this check, we AV when loading a fragile ngened image.

Commit:
Merge pull request #23858 from sandreenko/fixOptPrepareTreeToReplacement

Clean optVNConstantPropOnTree.

briansull committed:
Merge pull request #23876 from briansull/jit-dump

Improved JitDump

luhenry committed:
Make GC.GetGCMemoryInfo public (#23779)

* Make GC.GetGCMemoryInfo public

This is to be used to allow users to optimize memory consumption based
on what's available to the GC and on the system.

This is based on https://github.com/dotnet/corefx/issues/34631

* Address reviews

- Pass GCMemoryInfo struct to native in place of many arguments
- Make GCMemoryInfo.HeapSize and GCMemoryInfo.Fragmentation long in place of IntPtr

* Address reviews

- Mark GCMemoryInfo readonly

* Rearrange fields and match managed/native names

* Fix compilation

* Use getter shorthand

* Address API Design Review

https://github.com/dotnet/corefx/issues/34631#issuecomment-481358549

* Fix comments

* Fix comments

* Do everything in managed

* Address review

 - Remove unecessary [StructLayout.Sequential]
 - Remove "_" prefix for parameters variables

filipnavara committed:
Move TimerQueue.TickCount Windows implementation to shared CoreLib (#23838)

EgorBo committed:
Move AssemblyName to shared (#23737)

janvorli committed:
Merge pull request #23824 from janvorli/remove_unix_cpu_groups_emulation

Remove Unix CPU groups emulation

sdmaclea committed:
Contextual reflection (#23740)

* Add ContextualReflection APIs

Add ContextualReflection APIs approved in dotnet/corefx#36236
Fix issue #22213

* SetParentAssembly even when IsCollectible()
* ContextualReflection tests

* PR Feedback
* Add more usage tests

Add using statement tests
Add bad usage tests including Assert.Throws<>

* Only initialize on set
* Add XML API comments
* Unify VerifyIsolation
* Fix unused expectedAssembly
* Remove ContextualReflectionScope throw
* Clean up TestResolveMissingAssembly et. al
* Remove unused QCall::AppDomainHandle
* Remove AppDomainBaseObject
* Pass AssemblyLoadContext as managed object to native
* Fix AssemblyLoadContextBaseObject packing

* AssemblyLoadContext backing stores

Use explicit backing stores for events and properties

* Remove StaticAsyncLocalCurrentContextualReflectionContext
* Remove PermissionSetObject

AaronRobinsonMSFT committed:
Remove overrelease in test (#23887)

jkoritzinsky committed:
If a delegate type is a WinRT delegate or a WinRT-projected delegate, allow default marshalling. (#23877)

sywhang committed:
Renaming 'failure' in counter tests (#23891)

dotnet-maestro[bot] committed:
[master] Update dependencies from dotnet/arcade (#23834)

* Update dependencies from https://github.com/dotnet/arcade build 20190409.1

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19209.1
- Microsoft.DotNet.Helix.Sdk - 2.0.0-beta.19209.1

* Update dependencies from https://github.com/dotnet/arcade build 20190409.2

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19209.2
- Microsoft.DotNet.Helix.Sdk - 2.0.0-beta.19209.2

* Update dependencies from https://github.com/dotnet/arcade build 20190410.7

- Microsoft.DotNet.Arcade.Sdk - 1.0.0-beta.19210.7
- Microsoft.DotNet.Helix.Sdk - 2.0.0-beta.19210.7

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/core-setup build 20190410.11 (#23894)

- Microsoft.NETCore.App - 3.0.0-preview5-27610-11

safern committed:
Merge remote-tracking branch 'dotnet/master' into merge-master-nullable

safern committed:
Fix build errors due to merge conflicts

safern commented:
Due to some files that were changed in master that didn't have the nullability enabled I had to fix some build errors due to missing annotations on new stuff added, could someone take a look at the last commit as well?

stephentoub commented on src/System.Private.CoreLib/shared/System/Runtime/Loader/AssemblyLoadContext.cs:
```diff
@@ -508,10 +508,10 @@ public ContextualReflectionScope EnterContextualReflection()
         {
             return activating != null ?
-                GetLoadContext(activating).EnterContextualReflection() :
+                GetLoadContext(activating)?.EnterContextualReflection() ?? default :
```
This is changing the implementation.  Is there actually a risk of a null dereference here?

safern replied:
Yes... `GetLoadContext` can actually return null if `activating` is not a `RuntimeAssembly` https://github.com/dotnet/coreclr/blob/206f9680303134731862a35338618f2cbd225651/src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs#L254

That is why I thought of guarding it, but actually , you're right, we shouldn't change the implementaiton, just maybe open an issue since that is the same case now, it could cause a null reference exception, especially since it is a public API and any implementation of `Assembly` could be passed down and just return `null`... I could add a TODO comment and open an issue to discuss this further.

stephentoub replied:
> just maybe open an issue since that is the same case now

Let's do that.  Please open an issue, `!` away the warning here, and tag this with `// TODO: https://issue/You/Opened`.  Thanks.

safern replied:
Sounds good. Thanks for taking a look!

stephentoub commented:
> This time there were some merge conflicts in AssemblyName, AssemblyLoadContext, TimerQueue, RuntimeAssembly and Environment.Win32 -- looking at those files in the merge commit would be appreciated to just make sure it looks good and that I didn't miss anything during the merge conflicts.

How significantly did they change?  You could also consider just accepting the versions from master and we can redo the annotations on those few.

safern commented:
> How significantly did they change? You could also consider just accepting the versions from master and we can redo the annotations on those few. That might be easier / less risky.

Actually they didn't change much, it was just figuring out the annotations that we added that were missing in master. It was mostly because we moved stuff around into shared partition in master, so splitted into a shared file and a CoreCLR one. It was similar to the previous conflicts I had with `Delegate`, just matching whatever we annotated in the right set of files.

safern committed:
PR Feedback
